nodeup/pkg/model/kubeproxy.go

```diff
@@ -0,0 +1,220 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package model
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/kops/pkg/flagbuilder"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/nodeup/nodetasks"
+)
+
+// KubeProxyBuilder installs kube-proxy
+type KubeProxyBuilder struct {
+	*NodeupModelContext
+}
+
+var _ fi.ModelBuilder = &KubeAPIServerBuilder{}
+
+func (b *KubeProxyBuilder) Build(c *fi.ModelBuilderContext) error {
+	if b.IsMaster {
+		// If this is a master that is not isolated, run it as a normal node also (start kube-proxy etc)
+		// This lets e.g. daemonset pods communicate with other pods in the system
+		if fi.BoolValue(b.Cluster.Spec.IsolateMasters) {
+			glog.V(2).Infof("Running on Master with IsolateMaster=true; skipping kube-proxy installation")
+			return nil
+		}
+	}
+
+	{
+		pod, err := b.buildPod()
+		if err != nil {
+			return fmt.Errorf("error building kube-proxy manifest: %v", err)
+		}
+
+		manifest, err := ToVersionedYaml(pod)
+		if err != nil {
+			return fmt.Errorf("error marshalling manifest to yaml: %v", err)
+		}
+
+		t := &nodetasks.File{
+			Path:     "/etc/kubernetes/manifests/kube-proxy.manifest",
+			Contents: fi.NewBytesResource(manifest),
+			Type:     nodetasks.FileType_File,
+		}
+		c.AddTask(t)
+	}
+
+	// Add kubeconfig
+	{
+		kubeconfig, err := b.BuildPKIKubeconfig("kube-proxy")
+		if err != nil {
+			return err
+		}
+		t := &nodetasks.File{
+			Path:     "/var/lib/kube-proxy/kubeconfig",
+			Contents: fi.NewStringResource(kubeconfig),
+			Type:     nodetasks.FileType_File,
+			Mode:     s("0400"),
+		}
+		c.AddTask(t)
+	}
+
+	// Touch log file, so that docker doesn't create a directory instead
+	{
+		t := &nodetasks.File{
+			Path:        "/var/log/kube-proxy.log",
+			Contents:    fi.NewStringResource(""),
+			Type:        nodetasks.FileType_File,
+			Mode:        s("0400"),
+			IfNotExists: true,
+		}
+		c.AddTask(t)
+	}
+
+	return nil
+}
+
+func (b *KubeProxyBuilder) buildPod() (*v1.Pod, error) {
+	c := b.Cluster.Spec.KubeProxy
+	if c == nil {
+		return nil, fmt.Errorf("KubeProxy not configured")
+	}
+
+	if c.Master == "" {
+		if b.IsMaster {
+			c.Master = "https://127.0.0.1"
+		} else {
+			c.Master = "https://" + b.Cluster.Spec.MasterInternalName
+		}
+	}
+
+	cpuRequest, err := resource.ParseQuantity(c.CPURequest)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing CPURequest=%q", c.CPURequest)
+	}
+
+	flags, err := flagbuilder.BuildFlags(c)
+	if err != nil {
+		return nil, fmt.Errorf("error building kubeproxy flags: %v", err)
+	}
+
+	image := c.Image
+	cmd := "echo -998 > /proc/$$$/oom_score_adj && kube-proxy --kubeconfig=/var/lib/kube-proxy/kubeconfig --resource-container = \"\" "
+	cmd += flags
+	// TODO: tee or similar so we get logs in kubectl logs
+	cmd += " 1>>/var/log/kube-proxy.log 2>&1"
```

Comment:
I can check if we have `tee` in the image, but otherwise I don't know of an easy way